tpmutil/mssim/mssim.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2018, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mssim implements the Microsoft simulator TPM2 Transmission Interface
+//
+// The Microsoft simulator TPM Command Transmission Interface (TCTI) is a
+// remote procedure interface donated to the TPM2 Specification by Microsoft.
+// Its primary implementation is the tpm_server maintained by IBM.
+//
+// https://sourceforge.net/projects/ibmswtpm2/
+//
+// This package implements client code to communicate with server code described
+// in the document "TPM2 Specification Part 4: Supporting Routines – Code"
+//
+// https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-4-Supporting-Routines-01.38-code.pdf
+package mssim
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+)
+
+// Constants defined in "D.3.2. Typedefs and Defines"
+const (
+	tpmSignalPowerOn  uint32 = 1
+	tpmSignalPowerOff uint32 = 2
+	tpmSendCommand    uint32 = 8
+	tpmSignalNVOn     uint32 = 11
+	tpmSessionEnd     uint32 = 20
+)
+
+// Config holds configuration parameters for connecting to the simluator.
+type Config struct {
+	// Addresses of the command and platform handlers.
+	//
+	// Defaults to port 2321 and 2322 on localhost.
+	CommandAddress  string
+	PlatformAddress string
+}
+
+// Open creates connections to the simulator's command and platform ports and
+// power cycles the simulator to initialize it.
+func Open(ctx context.Context, config Config) (*Conn, error) {
+	cmdAddr := config.CommandAddress
+	if cmdAddr == "" {
+		cmdAddr = "127.0.0.1:2321"
+	}
+
+	platformAddr := config.PlatformAddress
+	if platformAddr == "" {
+		platformAddr = "127.0.0.1:2322"
+	}
+
+	dial := (&net.Dialer{}).DialContext
```

Comment:
hmm, let's use net.Dial and drop the ctx argument from Open
if there's any demand, we can add a Timeout field to Config and use net.DialTimeout

Reply:
done